File: string/Easy/ValidParentheses/ValidParentheses.py
 # Copyrights to venkys.io
# For more information, visit https://venkys.io
def isValidParanthesis(string: str) -> bool:
    stack = list()

    # Loop through each character in the input string
    for char in string:
        # If the character is an opening parenthesis, add it to the stack
        if char == "(" or char == "{" or char == "[":
            stack.append(char)
        else:
            # If the stack is empty and we encounter a closing parenthesis, return False
            if not stack:
                return False
            # Pop the top element from the stack
            top = stack.pop()
            # Check if the current closing parenthesis matches the corresponding opening parenthesis
            if char == ")" and top != "(":
                return False
            if char == "}" and top != "{":
                return False
            if char == "]" and top != "[":
                return False

    # If the stack is empty at the end, the parentheses are balanced
    return len(stack) == 0

# Get user input for the string
user_input = input("Enter a string with parentheses: ")
result = isValidParanthesis(user_input)

# Display the result
if result:
    print("The parentheses are valid.")
else:
    print("The parentheses are not valid.")

File: string/Easy/ValidParentheses/ValidParentheses.java
// Copyrights to venkys.io
// For more information, visit https://venkys.io
import java.util.Stack;
import java.util.Scanner;

public class ValidParentheses {

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        // Get user input for the string
        System.out.print("Enter a string with parentheses: ");
        String userInput = scanner.nextLine();

        // Call the isValid method with the user input
        boolean result = isValid(userInput);

        // Display the result
        if (result) {
            System.out.println("The parentheses are valid.");
        } else {
            System.out.println("The parentheses are not valid.");
        }

        // Close the scanner
        scanner.close();
    }

    // Method to check if parentheses are valid
    public static boolean isValid(String s) {
        Stack<Character> stack = new Stack<Character>();

        // Loop through each character in the input string
        for (char c : s.toCharArray()) {
            // If the character is an opening parenthesis, push the corresponding closing parenthesis to the stack
            if (c == '(')
                stack.push(')');
            else if (c == '{')
                stack.push('}');
            else if (c == '[')
                stack.push(']');
            else if (stack.isEmpty() || stack.pop() != c)
                return false;
        }

        // If the stack is empty at the end, the parentheses are balanced
        return stack.isEmpty();
    }
}
